### DISCUSSION TITLE
Important changes coming in Quarkus 3.14

### DISCUSSION THREAD

**Discussion by gsmet** [quarkiverse/quarkiverse#228] [General]
Hi @quarkiverse/quarkiverse-members ,

For Quarkus 3.14, a new extension annotation processor has been developed.

This will probably require some changes in your extension once you update to 3.14.

I have described the changes in details in the 3.14 migration guide: https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.14#for-extension-developers .

If you have any questions or encounter any issues, you can ping me anytime, I'm here to help.

Thanks.

-- 
Guillaume

**Comment by gastaldi** [comment]
@gsmet are those changes automatically applied when a new extension is created using `quarkus create extension`?

**Reply by gsmet** [reply]
Not at the moment, but I created https://github.com/quarkusio/quarkus/issues/42401 yesterday and I will fix it before CR1.

**Comment by turing85** [comment]
Hi!

Thanks Guillaume for the nice guide!

in the migration guide, section "publishing the config documentation" 
[1], it may be worth mentioning that the change only affects the "docs" 
submodule. If one takes a close look at the patch, then one can see 
this. But I was confused for a minute. WDYT?

[1]: 
https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.14#publishing-the-config-documentation

**Reply by gsmet** [reply]
I added some clarifications.

**Comment by ppalaga** [comment]
Is there a Quarkiverse extension repo, where the migration was done already?

**Reply by gastaldi** [reply]
I think 3.14 needs to be released for that to happen?

**Reply by ppalaga** [reply]
It would be interesting to see how the generated adoc looks like and whether there is any way to customize it?

**Reply by gsmet** [reply]
It looks exactly the same.
The point was to reimplement the infra, the output stayed mostly the same (but is a lot easier to update and improve now).

**Reply by gsmet** [reply]
@ppalaga I plan to update Quarkus GitHub App once CR1 is released. I'll update the guide then.

**Reply by ppalaga** [reply]
OK, thanks, if the AsciiDoc output looks the same, then we'd prefer to keep using our own plugin to transform the JSON metadata to AsciiDoc. 

>  the extension annotation processor builds a model that is in each module

So the JSON metadata files are now available per module, right? Where exactly?

**Reply by maxandersen** [reply]
@ppalaga can you remind me what it is you need changing in the output ? its not good IMO if extension starts diverging on this as we are also heading towards doing similar to what camel does of  generating artifacts IDE's uses. so lets make sure we get the needs captured.

**Reply by gsmet** [reply]
Same as @maxandersen . Except if you have oddly specific requirements, it would be better to improve the centralized plugin, rather than having people maintaining specific machinery.

And if you have oddly specific requirements, probably a good idea to discuss it to see if we can make the new stuff configurable for you.

We used to be very shy of updating this code as it was hard to evolved and very brittle. The new code is a lot more clean and flexible and the model generation is split from the output generation so we should be able to tweak the output a lot more easily.

**Reply by ppalaga** [reply]
There are several issues with the current state of the GenerateAsciidocMojo and the annotation processor. Some of them are easy to solve within the current design, some of them are harder:

- Change the qute template so that there is less unused space - see https://github.com/quarkiverse/antora-ui-quarkiverse/issues/56 and see how  we already render in Quarkus CXF: https://docs.quarkiverse.io/quarkus-cxf/dev/reference/extensions/quarkus-cxf.html#quarkus-cxf_quarkus-cxf-codegen-wsdl2java-includes
- Render `@since` version if available in JavaDoc. The metadata is correctly passed to yaml. Again, only the Qute template would have to get adjusted. This how we currently render it in Quarkus CXF: 
![image](https://github.com/user-attachments/assets/d21d937a-39a6-4548-8103-c90af288b126)
- Camel Antora UI required some special AsciiDoc markup, but I do not remember the details. Being able to override the default templates would perhaps solve the issue. 
- For Camel Quarkus and Quarkus CXF, we have custom plugins in place, that assemble the whole extension reference page (each extension has one) - see e.g. https://docs.quarkiverse.io/quarkus-cxf/dev/reference/extensions/quarkus-cxf-rt-features-metrics.html or https://camel.apache.org/camel-quarkus/3.8.x/reference/extensions/aws2-s3.html Note that in those pages the config section is optional (because some extensions do not have Quarkus config). The other sections are generated from extension metadata and some adoc chunks stored in src/adoc, such as usage.adoc, limitations.adoc, etc. We could in theory include the config section generated by GenerateAsciidocMojo, which we also did in the past, but that kind of solution was complex and hard to maintain and debug. Therefore we switched to our own custom plugins that assemble the whole reference page. I wonder if Quarkus team would be ready to widen the scope of the plugin to output the whole reference page?
- Our current setup does not require to rebuild the docs module when something in an extension was changed. It is enough to rebuild the extension. The current Quarkus solution requires rebuilding the docs module. I wonder if you have some CI in place that amends the generated changes when a dev forgets to rebuild the docs module?

**Reply by gsmet** [reply]
I don't see anything that would be problematic with the new design.

Maybe let's have this discussion in the Quarkus tracker instead of here, given all the Quarkiverse members get notified?

**Reply by gsmet** [reply]
I started a discussion here to address your feedback: https://github.com/quarkusio/quarkus/discussions/42476

**Comment by turing85** [comment]
Speaking of which: is 2024-08-14 still the ETA for 3.14.0? And can we 
expect a "CR1" before the final release?

**Reply by gsmet** [reply]
https://github.com/quarkusio/quarkus/wiki/Release-Planning#314-proposal

**Comment by metacosm** [comment]
I think it would be interesting to add the `@ConfigMapping` compatibility level with different Quarkus version because the approach chosen by developers might depend on their backward compatibility constraints with older Quarkus versions. In my case, for example, I'm wondering if it's better to rewrite the configuration sub-system using only `@ConfigMapping` or go the mixed-approach way.

**Reply by gsmet** [reply]
What do you mean by `mixed approach`?

Historically, for extensions, you were doing:

```java
@ConfigRoot(name = "github-app", phase = ConfigPhase.RUN_TIME)
public class GitHubAppRuntimeConfig {
```

When switching to the new and improved `@ConfigMapping` for extensions, you need to use:

```java
@ConfigRoot(phase = ConfigPhase.RUN_TIME)
@ConfigMapping(prefix = "quarkus.github-app")
public interface GitHubAppRuntimeConfig {
```

3.8 supports `@ConfigMapping` for extensions but we fixed some bugs in corner cases along the way so if you have a complex config, you might have some issues.

**Reply by michael-simons** [reply]
There is no such thing as "Mixed approach" as far as I understand it @metacosm 
You need both annotations.
And when you use the small rye one (`@ConfigMapping`), the annotation processor ensures you cannot use `name` and or `prefix` of `@ConfigRoot` anymore. However, the phases of `@ConfigRoot` are still required.

**Comment by gsmet** [comment]
If some of you are considering moving to `@ConfigMapping` (it's not mandatory but it can be a good opportunity to do it), you can have a look at what I did here: https://github.com/quarkiverse/quarkus-github-app/pull/645/files#diff-e57f34825aa8598a1397b16e4cf50b3a67672d287ea6e4946203ded8e5f5d484 .

One very important thing is that the prefix has to be moved to the `@ConfigMapping` annotation and that you need to prepend it with `quarkus.`.

**Comment by michael-simons** [comment]
What's the point of the change? Couldn't find it in the links shared here.
As 3rd party contributor, the move feels pointless. Care to elaborate now using Interfaces and proxies?

**Reply by michael-simons** [reply]
Not a trivial change https://github.com/quarkiverse/quarkus-neo4j/pull/284/commits/27dd2d90819d5fd3af6652eddbb45fe473fe37fe

**Reply by radcortez** [reply]
We have been discussing the `class` vs `interface` for some time now, including in the Jakarta Config group, with the following conclusions: https://github.com/jakartaee/config/discussions/122

In Quarkus, these are not proxies. These are just generated implementations during build time. It will still take a while for us to drop the class support, but our recommendation is to use interfaces since they now support a lot more features than class. Here are a few: https://smallrye.io/smallrye-config/3.9.1/config/mappings/

If you find any issues, I'm happy to help. Sorry for any inconvenience.

**Reply by gsmet** [reply]
@michael-simons well, the change is actually trivial if you don't move from the traditional class approach to the new interface approach. I don't think I mentioned that at all in what I wrote. But if I did, please point it to me as it's definitely not mandatory.

**Reply by michael-simons** [reply]
Thanks for chiming in, @radcortez 

As an "end-developer" (similar to end-user) experience, Springs type safe way just works best, and I can use it both with classes or records, whatever I like. 
The link has little discussion about the why's (i.e. what's the issue with default ctor and such), but alas.

I personally find the choice of targeting JDK 11 the most irritating thing, records would have been perfect, but than I don't follow the Jakarta politics about which version targets which other that closely, so I guess there are reasons.

Thanks again, though.

**Reply by michael-simons** [reply]
@gsmet I understood that it is an optional change, and I appreciate that you have the alternative in the migration guide. However, I absolutely hate things like that luring around just to bite me when I least expect it. Sure, I could have said something here already, but even breaking the build in a minor due to this change, is a bit of a surprise move, and I would prefer doing so in a major. So, I feel safer to migrate now, than at some point in the future, where I might not have the cycles to work on it.

(now off to 2 more extension…)

**Reply by michael-simons** [reply]
@radcortez There's that point that primitive values have already defaults and I read it like you think that it's a con argument. I would think it's pro, as most developers know the primitives default values, they are the same everywhere, and I found it pretty convenient.

**Reply by gsmet** [reply]
I'm not sure I understand your perspective: you have one compiler argument to add to unbreak your build if you're using legacy config roots (and then the Maven plugin to add but it doesn't break your build). And your build isn't broken if you already use config mappings.

It should be 5 mn of work, especially since I took the time to thoroughly document the changes to make. You wanted to also migrate to the new `@ConfigMapping` and that's fine but it's not really fair to put that on this change. It has nothing to do with it and it was your decision to do it (and that's great, don't get me wrong).

I'm not sure having to wait for the next major for this would have been a better bet given how many things it was blocking.
We did changes that were a lot more annoying in the extension world.

**Reply by gsmet** [reply]
Also in case it wasn't clear, this wasn't done to have fun (believe me, it wasn't ;)), it was a highly needed change and a source of technical debt that was in the way of several initiatives.
We don't do these kinds of changes lightly.

**Reply by michael-simons** [reply]
>  using legacy config roots 

That is the key point of your statement. They will not be legacy forever, they will vanish at some point in the future idk yet. I have to take care of this either way. I could ignore this by mitigating right now, or bite it and migrate now. I would be annoyed either way, but if Quarkus changed it on a major, and yes, carrying the legacy around for one major, less so. :)

**Reply by gsmet** [reply]
They won't but they have been legacy for quite a long time. What I want to emphasize is that it's orthogonal and I really don't have an issue with people going the easy way of adding the compiler argument - that's why I documented this approach. We won't break it any time soon - especially since the core is still full of legacy config roots :).

**Reply by michael-simons** [reply]
That's good to know, but than I hope for sure that they don't keep standing in the way of other initiatives then :)

**Reply by radcortez** [reply]
> I personally find the choice of targeting JDK 11 the most irritating thing, records would have been perfect, but than I don't follow the Jakarta politics about which version targets which other that closely, so I guess there are reasons.

Well yes, we had to go for the minimum supported version. While records could be supported, we had to come up with something for pre record support.

**Reply by radcortez** [reply]
> @radcortez There's that point that primitive values have already defaults and I read it like you think that it's a con argument. I would think it's pro, as most developers know the primitives default values, they are the same everywhere, and I found it pretty convenient.

Con, in a sense, you would require extra specifications to detail how things should behave. Nothing is wrong with that piece, but the group understood that the configuration may be set by an operator, who may not be a developer and not necessarily know the differences between primitives and wrapper types.

**Comment by michael-simons** [comment]
What about Quarkus unit tests, I have 

```java
	@RegisterExtension
	static final QuarkusUnitTest unitTest = new QuarkusUnitTest()
		.setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
			.addClass(SomeClass.class)
			.addClass(SomeOtherClass.class))
		.overrideConfigKey("org.neo4j.ogm.base-packages", "org.neo4j.ogm.quarkus.test.domain")
		.overrideConfigKey("org.neo4j.ogm.use-native-types", "true")
		.overrideConfigKey("org.neo4j.ogm.use-strict-querying", "true")
		.overrideConfigKey("org.neo4j.ogm.database", "aDatabase");
```

and now that tests yells at me 

```
java.lang.RuntimeException: io.smallrye.config.ConfigValidationException: Configuration validation failed:
	SRCFG00050: org.neo4j.ogm.use-native-types in PropertiesConfigSource[source=file:/tempfile/application.properties]:6 does not map to any root
```

with `@ConfigMapping`? Can't find a hint on that.

**Reply by michael-simons** [reply]
That one here: https://github.com/neo4j/neo4j-ogm-quarkus/blob/dependabot/maven/quarkus.version-3.14.0/deployment/src/test/java/org/neo4j/ogm/quarkus/test/ConfigurationTest.java

**Reply by gsmet** [reply]
I'm not sure how it's related to this change given the properties don't look like Quarkus properties (they don't have the `quarkus` prefix AFAICS)? Maybe this discussion belongs elsewhere?

**Reply by michael-simons** [reply]
No, they are not prefixed with Quarkus, but the build is broken nevertheless without the workaround or the migration.
The reason they are not prefixed is that the extension does not live in the quarkiverse. I pointed to the dependabot branch, with the interim migration.

**Reply by michael-simons** [reply]
If I do prefix them all with quarkus, the migration works, but that cannot be the solution.

**Reply by gsmet** [reply]
Again, probably not the right forum for this particular issue, it has nothing to do with what I introduced here. Thanks!

**Reply by michael-simons** [reply]
It has. 

But alas, where do you think asking that migration question is appropriate?

**Reply by michael-simons** [reply]
Thing is, a config mapping such as

```
@ConfigRoot(prefix = "org.neo4j", name = "ogm", phase = ConfigPhase.RUN_TIME)
```

cannot be migrated (or created neither) with `@ConfigMapping`, as the new machinery is pretty much focused on the `quarkus` prefix.  . 

You can berate me that this "forum" is the wrong place, or you could have created a ticket in Quarkus. I think it's the right place, as I feel you brought this up to discuss this change in a broader audience, but I am wrong it seems. I just wanted to help here. 

Fwiw, I can't even use the mitigation, as the annotation processor cannot be configured to not spit out a warning, thus breaking my build ( `-Werror`), which I really don't wanna remove, so that I am aware about general things. No, `Xlint:-processing` does not help, as that only suppresses missing processors.

**Reply by radcortez** [reply]
`@ConfigMapping` is a bit stricter in some validations, namely reporting configuration that is not mapped in the interface. The goal is to let the user know that either he has a configuration that shouldn't be there or he forgot to map that configuration.

In this case, the configuration is indeed valid, but for runtime only, and not build-time, so the build-time mapping reports that as a problem. We usually assume that extensions share the `quarkus` namespace, and there is a mapping ignore rule for configurations in the `quarkus` namespace.

It is possible to add additional ignore rules, by providing a `SmallRyeConfigBuilderCustomizer` and providing the ignore pattern in `io.smallrye.config.SmallRyeConfigBuilder#withMappingIgnore`: https://smallrye.io/smallrye-config/3.9.1/config/customizer/

Again, sorry for the inconvenience. If you wish, I'm happy to provide the necessary PRs to provide the adjustments.

**Reply by michael-simons** [reply]
How do I add such a customiser with a unit test like this

```java
static QuarkusUnitTest test = new QuarkusUnitTest()
		.withConfigurationResource("application.properties")
		.overrideConfigKey("org.neo4j.migrations.locations-to-scan", "classpath:neo4j/secondary-migrations")
		.overrideConfigKey("org.neo4j.migrations.external-locations", p.getParent().toUri().toString())
		.withApplicationRoot(archive -> {
			archive.addAsResource("neo4j/migrations");
			archive.addAsResource("neo4j/secondary-migrations");
		});
```

or the one above?

**Reply by michael-simons** [reply]
Oh, this is done globally, I see :/

**Reply by radcortez** [reply]
Yes, register it as a `ServiceLoader` file in the runtime module. You want this for user applications, not just tests.

**Reply by michael-simons** [reply]
Does not work, though… https://github.com/neo4j/neo4j-ogm-quarkus/tree/try-new-config

I see my customiser being called, the pattern ignored, but not only for validation, but for the whole config, and the test fails (`ConfigurationTest`) 

Happy to see a PR on that branch, really.

**Reply by radcortez** [reply]
Ok, let me check. I may have forgotten something.

**Reply by radcortez** [reply]
It works. Issue is in the mapping prefix. Was `quarkus.neo4j.ogm`, but it needs to be `org.neo4j.ogm`, since this is the config namespace you are using in the test, unless you also want to migrate that :)

https://github.com/neo4j/neo4j-ogm-quarkus/pull/372

**Reply by michael-simons** [reply]
That's brilliant, thank you so much @radcortez This is obviously my fault (Was trying things out, because those kinda migrations / customisations are not documented as a whole).

See, and that's my reason why I thing that discussions belongs here, @gsmet  

I was initially pulled in into this thread with reason of the new annotation processor.
Ofc I can do Tickets somewhere else and I have a proven track record of doing this (and fixing them my own many times). But here we already had a discussion, and from my PoV it was the exact right spot.
For sure, my implied feedback is not positive, but take it nevertheless sent with good intention. The quarkus team makes a move, that requires a lot of changes either by following along, or swallowing some changes in config with unnecessary consequences (the warning sent to the compiler that can't be disabled), and maybe not everyone can understand or see the reason behind it immediately. 
Also take note that I not once asked to change anything, but just trying to figure out how to *migrate* proper.
Thanks.

**Reply by gastaldi** [reply]
> See, and that's my reason why I thing that discussions belongs here, @gastaldi

I think you mean @gsmet ? Anyway, glad it got all sorted out 😉

**Reply by michael-simons** [reply]
D'oh. Sorry.

**Comment by michael-simons** [comment]
For people coming after me:

Here's the commit in an application with a straight forward migration to the new config processor:

https://github.com/quarkiverse/quarkus-neo4j/commit/585d499bd813c9b8232ea04df6caab8a9b81d48e

And here's an another extension that does not live in the Quarkiverse and hence does not use the `quarkus.` prefix in its configuration properties. It requires a [Smallrye config customizer](https://smallrye.io/smallrye-config/Main/config/customizer/) ignore the config patterns under the prefix, the same way it does for properties prefixed with `quarkus.*`. 
Take note that the pattern must end in `whatever.**` (two asterisks):

https://github.com/neo4j/neo4j-ogm-quarkus/commit/fe8687626a265ffa0f6ed5cc9d7f5f7673683976

Thanks a lot for helping me out with the latter, @radcortez I would not have found out this alone.

**Reply by gsmet** [reply]
Sorry but I will have to reiterate: what you point out is a migration to the new extension annotation processor + a migration to @ConfigMapping. 

The second part is not necessary and is orthogonal to what is required to migrate to 3.14. 

Let’s not create confusion here. Thanks.

**Reply by michael-simons** [reply]
Your change in Quarkus requires one of two actions. One is going down the deprecation path, one is the migration. Both have consequences, both are none trivial. Thanks for devaluing my positive comment that describes how to do one of them. 

I’m unsubscribed now from this and frankly, quite disappointed. Lesson learned: feedback is not appreciated here.
